src/main/java/com/aws/greengrass/ipc/GetClientDeviceAuthTokenOperationHandler.java

```diff
@@ -0,0 +1,128 @@
+    private static final String UNAUTHORIZED_ERROR = "Not Authorized";
+    private static final String MQTT_CREDENTIAL_TYPE = "mqtt";
+    private static final String NO_DEVICE_CREDENTIAL_ERROR = "Client device credential document is required";
+    private static final String NO_VALID_MQTT_CREDENTIAL_ERROR = "Valid mqtt credential document is required";
+    private final String serviceName;
+    private final AuthorizationHandler authorizationHandler;
+    private final SessionManager sessionManager;
+    private final Map<String, String> credentialMap = new HashMap<>();
+
+    /**
+     * Constructor.
+     *
+     * @param context              operation continuation handler
+     * @param sessionManager       session manager
+     * @param authorizationHandler authorization handler
+     */
+    public GetClientDeviceAuthTokenOperationHandler(
+            OperationContinuationHandlerContext context,
+            SessionManager sessionManager,
+            AuthorizationHandler authorizationHandler
+    ) {
+
+        super(context);
+        serviceName = context.getAuthenticationData().getIdentityLabel();
+        this.sessionManager = sessionManager;
+        this.authorizationHandler = authorizationHandler;
+    }
+
+    @SuppressWarnings({"PMD.AvoidCatchingGenericException", "PMD.PreserveStackTrace"})
+    @Override
+    public GetClientDeviceAuthTokenResponse handleRequest(GetClientDeviceAuthTokenRequest request) {
+        return translateExceptions(() -> {
+            try {
+                doAuthorizationForClientDevAuthToken();
+            } catch (AuthorizationException e) {
+                logger.atWarn().kv("error", e.getMessage()).kv(COMPONENT_NAME, serviceName).log(UNAUTHORIZED_ERROR);
+                throw new UnauthorizedError(e.getMessage());
+            }
+            Map<String, String> credentialMap = mapOfMqttCredential(request.getCredential());
+            try {
+                String sessionId = sessionManager.createSession(MQTT_CREDENTIAL_TYPE, credentialMap);
+                GetClientDeviceAuthTokenResponse response = new GetClientDeviceAuthTokenResponse();
+                return response.withClientDeviceAuthToken(sessionId);
+            } catch (Exception e) {
+                logger.atError().cause(e).log("Unable to get client device auth token from the session");
+                throw new ServiceError(
+                        "Getting client device auth token failed. Check Greengrass log for details.");
+            }
+        });
+    }
+
+    @SuppressWarnings("PMD.PreserveStackTrace")
+    private MQTTCredential validateMqttCredentials(CredentialDocument credentialDocument) {
+        if (credentialDocument == null) {
+            throw new InvalidArgumentsError(NO_DEVICE_CREDENTIAL_ERROR);
+        }
+        MQTTCredential mqttCredential = credentialDocument.getMqttCredential();
+        if (mqttCredential == null) {
+            throw new InvalidArgumentsError(NO_VALID_MQTT_CREDENTIAL_ERROR);
```

Comment:
I'd keep this basic. "Invalid device credentials"

Reply:
L91 gives `Client device credential document is required` and L95 gives `Valid mqtt credential document is required`. Are you suggesting `Invalid mqtt credentials`  for L95?

Reply:
I'm saying not to return detailed information to the caller about what failed. You can log extra information here, but just return `Invalid device credentials` to the caller.